### PR TITLE
feat: Switch to async clickhouse client in squash workflow

### PR DESCRIPTION
**Pull request by tomasfarias**
## Problem

Squash workflow doesn't use async `ClickHouseClient`, which means it would block any other workflows. Let's switch it!

## Changes

- [x] Move `clickhouse.py` to common utilities.
- [x] Updated squash workflow to use ClickHouseClient
- [x] Added support for `mutations_sync` when initializing a client which makes tests that run mutations less flaky.

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Updated unit tests

### CODE REVIEW

File: posthog/temporal/tests/test_clickhouse.py
Comment (tkaemming) [inline posthog/temporal/tests/test_clickhouse.py:1]: Should this be updated to reflect the move to `temporal.common.clickhouse` as well?

File: posthog/temporal/common/clickhouse.py [hunk 1]
```diff
@@ -49,12 +49,29 @@ def encode_clickhouse_data(data: typing.Any) -> bytes:
...
+                    quote_char = '"'  # JSON requires double quotes.
+                else:
+                    quote_char = "'"
+
+                encoded_data.append(
+                    f'"{str(key)}"'.encode("utf-8") + b":" + encode_clickhouse_data(value, quote_char=quote_char)
+                )
+
+            result = b"{" + b",".join(encoded_data) + b"}"
+            return result
```
Comment (tkaemming) [inline posthog/temporal/common/clickhouse.py:69]: A comment here about why we're building these strings this way would be helpful (it's not obvious to me, at least.)

File: posthog/temporal/batch_exports/squash_person_overrides.py [hunk 2]
```diff
@@ -352,18 +370,31 @@ async def select_persons_to_delete(inputs: QueryInputs) -> list[SerializablePers
...
+            UUID,
+            UUID,
+            parse_clickhouse_timestamp,
+            int,
+            parse_clickhouse_timestamp,
+        )
+        to_delete_rows = (
+            (schema[field_index](field_value) for field_index, field_value in enumerate(line.split("\t")))
+            for line in response.decode("utf-8").splitlines()
+        )
```
Comment (tkaemming) [inline posthog/temporal/batch_exports/squash_person_overrides.py:397]: Spicy!

File: posthog/temporal/batch_exports/squash_person_overrides.py [hunk 3]
```diff
@@ -437,17 +470,14 @@ async def squash_events_partition(inputs: QueryInputs) -> None:
...
         activity.logger.info("Executing squash query on partition %s", partition_id)
 
         parameters = {
             "partition_id": partition_id,
             "team_ids": inputs.team_ids,
-            # We pass this as a timestamp ourselves as clickhouse-driver will drop any microseconds from the datetime.
-            # This would cause the latest merge event to be ignored.
-            # See: https://github.com/mymarilyn/clickhouse-driver/issues/306
             "latest_created_at": latest_created_at,
         }
```
Comment (tkaemming) [inline posthog/temporal/batch_exports/squash_person_overrides.py:482]: It's not clear to me why the way `QueryInputs.last_created_at` is treated here changed, while this other usage didn't: https://github.com/PostHog/posthog/blob/270bff77be34dafc686055d484119d1df8e897b5/posthog/temporal/batch_exports/squash_person_overrides.py#L372-L381

File: posthog/temporal/batch_exports/squash_person_overrides.py [hunk 4]
```diff
@@ -579,6 +615,7 @@ class SquashPersonOverridesInputs:
     dictionary_name: str = "person_overrides_join_dict"
     last_n_months: int = 1
     dry_run: bool = True
+    wait_for_mutations: bool = False
```
Comment (tkaemming) [inline posthog/temporal/batch_exports/squash_person_overrides.py:618]: 👌

**Review by tkaemming — APPROVED** [review]

File: posthog/temporal/common/clickhouse.py [hunk 5]
```diff
@@ -138,13 +155,13 @@ def prepare_query(self, query: str, query_parameters: None | dict[str, typing.An
         Returns:
             The formatted query.
         """
-        if query_parameters:
-            format_parameters = {k: encode_clickhouse_data(v).decode("utf-8") for k, v in query_parameters.items()}
-        else:
-            format_parameters = {}
+        if query_parameters is None or not query_parameters:
```
[the remark this answers is not in the document]
Comment (tomasfarias) [reply posthog/temporal/common/clickhouse.py:158]: Yeah, I think I was being wary of the type checker who can be stingy if you don't explicitly check for `None`. But I've tested it and that doesn't seem the case here. I'll use just `not query_parameters`.

File: posthog/temporal/batch_exports/squash_person_overrides.py (hunk 3 above)
Comment (tomasfarias) [reply posthog/temporal/batch_exports/squash_person_overrides.py:482]: New client handles timestamps without losing precision, so we don't need to `timestamp()` it ourselves. Nothing is changing, mostly just pushing the operation until the last moment.

File: posthog/temporal/batch_exports/squash_person_overrides.py (hunk 2 above)
Comment (tomasfarias) [reply posthog/temporal/batch_exports/squash_person_overrides.py:397]: Yeah, all of this going away in second iteration...

File: posthog/temporal/tests/test_clickhouse.py
Comment (tomasfarias) [reply posthog/temporal/tests/test_clickhouse.py:1]: We don't have a `tests/common` or `tests/common/clickhouse` So I'd keep it here for now, but eventually yes.

File: posthog/temporal/common/clickhouse.py (hunk 1 above)
Comment (tomasfarias) [reply posthog/temporal/common/clickhouse.py]: I've extended the top comment explaining a bit more why we do JSON encoding for dictionaries and its limitations.

**Pull request merged**